### PR TITLE
GLOSA Device Update

### PR DESCRIPTION
**Pull request by Domsall**
Apart from many new Debug output commands, we added more options to the GLOSA device:

1. New option `device.glosa.use-queue` that uses an E2-Detector to calculate the queue length of vehicles in front of the traffic light, which is used to update the arrival time of the GLOSA-vehicle
2. We updated the calculation of the next traffic light's phase. The GLOSA-vehicle now not only calculates the time to the next phase switch, but to the phase when the vehicle arrives at the intersection (phase calculation is in a loop).
3. Added a switch time offset to the looped phase calculation. This makes GLOSA-vehicles drive a bit faster when they want to reach the beginning or end of a green phase. Result: Fewer stops and higher traffic flow. 
4. Updated one minor test change

### CODE REVIEW

**Comment by m-kro** [comment]
@Domsall Some meta tests still fail (probably due to the new option). Can you please run them and eventually confirm the new outcome (if you see nothing else wrong with them)?

**Comment by Domsall** [comment]
you are right, I did not know that the options can be written to an output and that tests exist for that. The new queue option is throwing the error, everything else is fine.

File: src/microsim/devices/MSDevice_GLOSA.cpp [hunk 1]
```diff
@@ -270,7 +278,18 @@ MSDevice_GLOSA::notifyMove(SUMOTrafficObject& /*tObject*/, double oldPos,
                 }
             }
             // calculate next Phase
-            nextSwitch += getTimeToNextSwitch(myNextTLSLink, currentPhaseGreen, currentPhaseStop, countOld); 
+            nextSwitch += getTimeToNextSwitch(myNextTLSLink, currentPhaseGreen, currentPhaseStop, countOld);
+            // For vehicles far away from the junction we add an offset
+            if (nextSwitch > 20. && switchOffset == 0) {
+                nextSwitch -= (double)switchOffset;
```
Comment (m-kro) [inline src/microsim/devices/MSDevice_GLOSA.cpp:284]: It does not seem useful to me to subtract 0?

File: src/microsim/devices/MSDevice_GLOSA.cpp [hunk 2]
```diff
@@ -118,37 +124,149 @@ MSDevice_GLOSA::notifyMove(SUMOTrafficObject& /*tObject*/, double oldPos,
...
+#ifdef DEBUG_QUEUE
+                if (DEBUG_COND) {
+                    std::cout << SIMTIME << " veh=" << myVeh.getID() << " Nextswitch(RedQueue)=" << nextSwitch << "\n";
+                }
+#endif
+            }
+        }
+
+        // Search for the next passable phase, maximum 10 Phases 
+        for (int countwhile = 1; countwhile < 10; countwhile++) {
```
[the remark this answers is not in the document]
Comment (Domsall) [reply src/microsim/devices/MSDevice_GLOSA.cpp:196]: You are right :) You could also use an "endless" while-loop, we wanted to have a safety-net

File: src/microsim/devices/MSDevice_GLOSA.cpp [hunk 3]
```diff
@@ -118,37 +124,149 @@ MSDevice_GLOSA::notifyMove(SUMOTrafficObject& /*tObject*/, double oldPos,
...
                     }
+#endif
+                }
+                adaptSpeed(myDistance, timeToJunction, nextSwitch + myAddSwitchTime, solved);
+                if (countwhile == 2 && myUseQueue) {
+                    // reset queue because following phases should not calculate with current queue
+                    nextSwitch -= QueueLength * factor + addition;
+                    timeToJunction -= additionaljunctiontime;
+                }
+                if (solved == true) {
```
Comment (m-kro) [inline src/microsim/devices/MSDevice_GLOSA.cpp:276]: no need for the equal operator as the variable is already a boolean

File: src/microsim/devices/MSDevice_GLOSA.cpp [hunk 4]
```diff
@@ -118,37 +124,149 @@ MSDevice_GLOSA::notifyMove(SUMOTrafficObject& /*tObject*/, double oldPos,
...
-                if (myMaxSpeedFactor > myVeh.getChosenSpeedFactor()) {
-                    const double vMax2 = vMax / myVeh.getChosenSpeedFactor() * myMaxSpeedFactor;
-                    const double timetoJunction2 = earliest_arrival(myDistance, vMax2);
-                    // reaching the signal at yellow might be sufficient
-                    const double yellowSlack = myVeh.getVehicleType().getParameter().getJMParam(SUMO_ATTR_JM_DRIVE_AFTER_YELLOW_TIME, 0);
+        if (myUseQueue) {
+		    // Detect queue length at tls
+		    double QueueLength = myNextTLSLink->getTLLogic()->getTLQueueLength(myNextTLSLink->getLaneBefore()->getID());
+            double greentime = 0;
+            double additionaljunctiontime = 0;
```
Comment (m-kro) [inline src/microsim/devices/MSDevice_GLOSA.cpp]: camel case variable naming style is preferred, see https://sumo.dlr.de/docs/Developer/CodeStyle.html#variable_names

Comment (Domsall) [reply src/microsim/devices/MSDevice_GLOSA.cpp]: Should I change it or will you?

File: src/microsim/devices/MSDevice_GLOSA.cpp (hunk 3 above)
Comment (Domsall) [reply src/microsim/devices/MSDevice_GLOSA.cpp:276]: Should I change it or will you?

File: src/microsim/output/MSE2Collector.cpp [hunk 5]
```diff
@@ -1553,15 +1553,15 @@ double
 MSE2Collector::getEstimateQueueLength() const {
 
     if (myVehicleInfos.empty()) {
-        return -1;
+        return 0;
```
Comment (m-kro) [inline src/microsim/output/MSE2Collector.cpp:1556]: Although this code part does not seem to be used much, it is an additional value to know immediately whether the queue length is due to absence of vehicles or they just didn't stop...

File: src/microsim/devices/MSDevice_GLOSA.cpp (hunk 3 above)
Comment (m-kro) [reply src/microsim/devices/MSDevice_GLOSA.cpp:276]: You can do it

File: src/microsim/devices/MSDevice_GLOSA.cpp (hunk 4 above)
Comment (m-kro) [reply src/microsim/devices/MSDevice_GLOSA.cpp]: You can do it

**Comment by Domsall** [comment]
All comments are fixed

**Comment by namdre** [comment]
@Domsall It appears that the new option `device.glosa.use-queue` only takes effect for `tlType="delay_based"`.  Do you have some results on how often the dynamic switching invalidates the GLOSA-advisory?

**Comment by Domsall** [comment]
> @Domsall It appears that the new option `device.glosa.use-queue` only takes effect for `tlType="delay_based"`. Do you have some results on how often the dynamic switching invalidates the GLOSA-advisory?

Yes... but we only used the `delay_based` type for the creation of the E2-detector. The traffic lights ran in a static loop.

**Comment by Domsall** [comment]
> also, would you care to add info on the new features at https://sumo.dlr.de/docs/Simulation/GLOSA.html ?

Its on my agenda

**Pull request merged**